### DISCUSSION TITLE
How can you import Viem-compatible contract types directly

### DISCUSSION THREAD

**Discussion by travs** [NomicFoundation/hardhat#4754] [Troubleshooting]
I was going to raise this as an issue, but I think it's just my not being able to find where to import contract instance types, since [this issue](https://github.com/NomicFoundation/hardhat/issues/4275) is already resolved.

Basically I just want to use the type of a contract that is deployed with something like:
```
await hre.viem.deployContract("TestToken")
```

Right now I have created a declaration file like the below, but it will get cumbersome to update/manage as the number of contract types grows. Not to mention that it requires hardhat artifacts to be created, which intellisense, linting, etc. won't work until the contracts are compiled.

```ts
// contractTypes.d.ts
import { GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";
import {TestToken$Type} from "../../artifacts/contracts/testing/TestToken.sol/TestToken";
export type TestToken = GetContractReturnType<TestToken$Type["abi"]>;
```

Is there any single place where I can just import the type?

**Comment by fvictorio** [comment]
The type lives in the `artifacts` directory and, as you said, it's auto-generated, so importing it directly is probably not ideal.

This workaround seems to work, but typescript won't compile until you compile your contracts. But at least it doesn't need you to use explicit paths:

```typescript
import { GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";
import { ArtifactsMap } from "hardhat/types/artifacts";

type TestToken = GetContractReturnType<ArtifactsMap["TestToken"]["abi"]>
```

A more involved alternative, but that will work even if you contracts are not compiled, is this:

```typescript
import { GetContractReturnType } from "@nomicfoundation/hardhat-viem/types";
import { ArtifactsMap } from "hardhat/types/artifacts";

type GetContractType<T extends string> = ArtifactsMap extends {[key in T]: any} ? GetContractReturnType<ArtifactsMap[T]["abi"]> : GetContractReturnType;
type TestToken = GetContractType<"TestToken">
```

I haven't tested this in depth, so please let me know if it works or doesn't!

**Comment by SebastienGllmt** [comment]
You mentioned `hre.viem` so I assume this is not in the context of your testing framework and not in an external tool,

but I made an update to the `hardhat-abi-exporter` plugin to export the right ABI to use directly in your frontend if that helps unblock you as well (you should be able to use the generated ts wrappers in your hardhat code as well)

It's not as comprehensive as the stuff that `hardhat-viem` generates and it also requires the task to be run for the intellisense to work, but hopefully it helps anyway

https://github.com/ItsNickBarry/hardhat-abi-exporter/pull/54
